// @internal/components-entry-legacy/index-browser.js
var Component = require("../../../runtime/components/Component");
var loader = require("@internal/loader");
var getComponentForEl = require("@internal/components-util").V_;

// expose legacy
require("@internal/components-registry")._j_ = exports;
exports.load = function (typeName) {
  return exports.defineWidget(loader(typeName));
};

// legacy api
exports.defineComponent = require("../../../runtime/components/legacy/defineComponent-legacy");
exports.defineWidget = require("@internal/components-define-widget-legacy");
exports.defineRenderer = require("../../../runtime/components/legacy/defineRenderer-legacy");
exports.makeRenderable = exports.renderable = require("../../../runtime/renderable");

// browser only
var Widget = exports.Widget = Component;
exports.getWidgetForEl = exports.get = function (elOrId) {
  var el = elOrId;

  if (typeof elOrId === "string") {
    el = document.getElementById(elOrId);
  }

  if (el && el.__widget) {
    return el.__widget;
  }

  return getComponentForEl(el);
};
exports.initWidgets = require("@internal/components-registry").W_;

// monkey patch Widget
if (Widget) {
  var WidgetProto = Widget.prototype;
  WidgetProto.setProps = function (newInput) {
    this._k_ = true;
    this._l_(newInput);
  };
  WidgetProto.rerender = function (newInput) {
    if (newInput) {
      this.setProps(newInput);
    }

    this.forceUpdate();
    this.update();
  };
}

var RenderResult = require("../../../runtime/RenderResult");

RenderResult.prototype.getWidget = function () {
  return this.getWidgets()[0];
};
RenderResult.prototype.getWidgets = function () {
  return RenderResult.prototype.getComponents.apply(this, arguments).filter(function (component) {
    return component.B_;
  });
};
