// @internal/components-define-widget-legacy/index-browser.js
"use strict";
/* jshint newcap:false */

var BaseState;
var BaseComponent;
var inherit;
var jQuery = require("../../../runtime/components/jquery");
var ready = require("../../../runtime/components/ready");

var complain;
var stateToJSONDef = {
  enumerable: false,
  value: function returnSelf() {
    return this;
  }
};
function noop() {}

module.exports = function defineWidget(def, renderer) {
  def = def.Widget || def;

  if (def.A_) {
    return def;
  }

  var ComponentClass = function () {};
  var proto;
  var legacyInit;

  if (typeof def === "function") {
    proto = def.prototype;
    legacyInit = def;
  } else if (typeof def === "object") {
    proto = def;
    legacyInit = def.init;
  } else {
    throw TypeError();
  }

  ComponentClass.prototype = proto;

  // We don't use the constructor provided by the user
  // since we don't invoke their constructor until
  // we have had a chance to do our own initialization.
  // Instead, we store their constructor in the "initComponent"
  // property and that method gets called later inside
  // init-components-browser.js
  function Component(id, doc) {
    BaseComponent.call(this, id, doc);
  }

  if (!proto.A_) {
    // Inherit from Component if they didn't already
    ComponentClass.prototype = Object.create(BaseComponent.prototype);
    for (var propName in proto) {
      if (proto.hasOwnProperty(propName)) {
        ComponentClass.prototype[propName] = proto[propName];
      }
    }
  }

  // The same prototype will be used by our constructor after
  // we he have set up the prototype chain using the inherit function
  proto = Component.prototype = ComponentClass.prototype;
  proto.B_ = true;

  proto.constructor = def.constructor = Component;

  Object.defineProperty(proto, "state", {
    get: function () {
      var raw = this.C_ && this.C_.D_;

      if (raw && !raw.toJSON) {
        Object.defineProperty(this.C_.D_, "toJSON", stateToJSONDef);
      }

      return raw;
    },
    set: function (newState) {
      newState = newState || {};
      this.setState(newState);
    }
  });

  Object.defineProperty(proto, "__document", {
    get: function () {
      return this.E_;
    }
  });

  Object.defineProperty(proto, "el", {
    get: function () {
      return this.F_;
    }
  });

  // get legacy methods

  proto.G_ = proto.onRender;
  Object.defineProperty(proto, "onRender", {
    get: noop,
    set: function (v) {
      proto.G_ = v;
    }
  });

  proto.H_ = proto.onUpdate;
  Object.defineProperty(proto, "onUpdate", {
    get: function () {
      return modernMountOrUpdate;
    },
    set: function (v) {
      proto.H_ = v;
    }
  });

  proto.I_ = proto.onDestroy;
  Object.defineProperty(proto, "onDestroy", {
    get: function () {
      return modernOnDestory;
    },
    set: function (v) {
      proto.I_ = v;
    }
  });

  proto.getWidget = proto.getComponent;
  proto.getWidgets = proto.getComponents;
  proto.onMount = modernMountOrUpdate;

  if (legacyInit) {
    proto.J_ = legacyInit;
  }

  // convert legacy to modern
  proto.K_ = proto.update;
  proto.update = function () {
    if (this.L_) {

      this.destroy = modernOnDestory;
      // eslint-disable-next-line no-constant-condition

      this.L_ = false;
    }
    this.M_ = true;
    if (this.F_) {
      this.onBeforeUpdate && this.onBeforeUpdate();
    }

    this.K_();
    this.M_ = false;
  };

  function modernMountOrUpdate() {
    var self = this;
    var el = this.N_["@_wbind"];
    var prevEl = this.F_;

    if (prevEl !== el) {
      this.F_ = el;

      if (prevEl) {
        this.onBeforeDestroy && this.onBeforeDestroy();
        this.I_ && this.I_();
        this.O_ = undefined;
        this.removeAllListeners();
      }

      if (el) {
        this.J_ && this.J_(this.widgetConfig || {});
        this.G_ && this.G_({ firstRender: true });
        this.O_ = legacyRender;
        el.__widget = this;
      }
    } else if (el) {
      if (prevEl) {
        this.H_ && this.H_();
      }

      if (this.P_) {
        this.G_ && this.G_({ firstRender: false });
      }
    }

    this.Q_ = this.R_;
    this.R_ = null;
    this.P_ = false;
  }

  function legacyRender() {
    if (!this.M_) {
      this.onBeforeUpdate && this.onBeforeUpdate();
    }

    this.P_ = true;
  }

  function modernOnDestory() {
    if (this.F_) {
      this.onBeforeDestroy && this.onBeforeDestroy();
      this.I_ && this.I_();
      this.F_ = null;
    }
  }

  // Set a flag on the constructor function to make it clear this is
  // a component so that we can short-circuit this work later
  Component.A_ = true;

  function State() {
    BaseState.apply(this, arguments);
  }
  inherit(State, BaseState);
  proto.S_ = State;

  jQuery.patchComponent(window.$, proto, true /* don't throw error until used if `$` is missing*/
  );

  ready.patchComponent(proto);

  if (!renderer) {
    renderer = ComponentClass.renderer || ComponentClass.prototype.renderer;
    if (renderer) {
      // Legacy support
      var createOut = renderer.createOut;
      if (typeof renderer !== "function") {
        var rendererObject = renderer;
        renderer = function (input, out) {
          var rendererFunc = rendererObject.renderer || rendererObject.render;
          rendererFunc(input, out);
        };
        renderer.createOut = createOut;
      }

      renderer.render = function (input) {
        var out = createOut();
        renderer(input, out);
        return out.end();
      };
    }
  }

  if (renderer) {
    // Add the rendering related methods as statics on the
    // new component constructor function
    Component.renderer = proto.T_ = renderer;
    Component.render = renderer.render;
    Component.renderSync = renderer.renderSync;
  }

  Component._isWidget = true;

  return Component;
};

BaseState = require("../../../runtime/components/State");
BaseComponent = require("../../../runtime/components/Component");
inherit = require("raptor-util/inherit");
